gadget/gadget.go

```diff
@@ -400,15 +416,76 @@ func validateVolume(name string, vol *Volume) error {
 			knownFsLabels[s.Label] = true
 		}
 
+		switch s.Role {
+		case SystemSeed:
+			if state.SystemSeed != nil {
+				return fmt.Errorf("cannot have more than one system-seed role")
+			}
+			state.SystemSeed = &vol.Structure[idx]
+		case SystemData:
+			if state.SystemData != nil {
+				return fmt.Errorf("cannot have more than one system-seed role")
+			}
+			state.SystemData = &vol.Structure[idx]
+		}
+
 		previousEnd = end
 	}
 
+	if err := ensureVolumeConsistency(state, constraints); err != nil {
+		return err
+	}
+
 	// sort by starting offset
 	sort.Sort(byStartOffset(structures))
 
 	return validateCrossVolumeStructure(structures, knownStructures)
 }
 
+func ensureVolumeConsistency(state *validationState, constraints *ModelConstraints) error {
+	if constraints == nil {
+		if state.SystemData != nil {
```

Comment:
Sorry for not proposing this earlier. I think it'd be a bit easier to read if the code was structured like this:
```
if state.SystemData == nil {
    return nil
}
if state.SystemSeed != nil {
    ....
} else {
    ....
}
```

Reply:
Refactored as proposed.